### PR TITLE
Prototech tests

### PR DESCRIPTION
**Pull request by godsflaw**
Work in progress PR with some invariants and a number of regressions.

### CODE REVIEW

File: src/ERC20Pool.sol [hunk 1]
```diff
@@ -290,6 +290,9 @@ contract ERC20Pool is FlashloanablePool, IERC20Pool {
         // revert if the dust amount was not exceeded, but round on the scale amount
         if (amountToAdd_ != 0 && amountToAdd_ < _bucketCollateralDust(index_)) revert DustAmountNotExceeded();
         amountToAdd_ = _roundToScale(amountToAdd_, _getArgUint256(COLLATERAL_SCALE));
+        if (amountToAdd_ < _bucketCollateralDust(index_)) {
+            require(false, "DustAmountNotExceeded");
+        }
```
Comment (godsflaw) [inline src/ERC20Pool.sol]: I'll remove this shortly, we were just testing if this would blow up because of doing the scaling after the first conditional.

File: tests/forge/regression/ERC721Pool/RegressionTestReservesERC721Pool.t.sol [hunk 2]
```diff
@@ -448,3 +448,19 @@ contract RegressionTestReserveERC721Pool is ReserveERC721PoolInvariants {
...
+contract RegressionTestReserveERC721PoolWith10BucketsIndex1500 is ReserveERC721PoolInvariants {
+
+    function setUp() public override {
+        // failures reproduced with 10 active buckets
+        vm.setEnv("NO_OF_BUCKETS", "10");
+        vm.setEnv("BUCKET_INDEX_ERC20", "1500");
+        super.setUp();
+    }
+
+    function test_regression_prototech_R6() external {
```
Comment (godsflaw) [inline tests/forge/regression/ERC721Pool/RegressionTestReservesERC721Pool.t.sol:498]: This entire regression may be a foundry fluke.  Invariant fuzzing failed here, but I cannot reproduce.

File: tests/forge/regression/ERC20Pool/RegressionTestBasicERC20Pool.t.sol [hunk 3]
```diff
@@ -466,3 +484,34 @@ contract RegressionTestBasicWith10BucketsERC20Pool is BasicERC20PoolInvariants {
...
+contract RegressionTestBasicWith10BucketsIndex6500ERC20Pool is BasicERC20PoolInvariants { 
+
+    function setUp() public override {
+        // failures reproduced with 10 active buckets
+        vm.setEnv("NO_OF_BUCKETS", "10");
+        vm.setEnv("BUCKET_INDEX_ERC20", "6500");
+        super.setUp();
+    }
+    
+    function test_regression_prototech_R7() external {
```
Comment (godsflaw) [inline tests/forge/regression/ERC20Pool/RegressionTestBasicERC20Pool.t.sol:815]: This is a real issue that appears to be throwing the same error I was chasing somewhere else.  Either the tests are incorrectly checking `require(lpBalanceAfterAction < lpBalanceBeforeAction, "LP balance should decrease")` or we have an issue where the LP balance is not updated despite being able to remove collateral.

Comment (grandizzy) [reply tests/forge/regression/ERC20Pool/RegressionTestBasicERC20Pool.t.sol:815]: that's a great catch :pray: running test with verbose `-vvv` reveals the fact actor removes collateral from pool for 0 LP redeemed
```
emit RemoveCollateral(claimer: Actor5: [0x46602Be05BF9087F45245071600fd65f9db70827], index: 6502, amount: 37232 [3.723e4], lpRedeemed: 0)
emit Transfer(from: 0x9Fd7552Da37D2D296CB4a84d507c35Dcc926220a, to: Actor5: [0x46602Be05BF9087F45245071600fd65f9db70827], value: 37232)
```

this is an issue and we covered similar cases with https://github.com/ajna-finance/contracts/pull/772 (by basically reverting with `InsufficientLP` if the LP awarded for assets added in pool is 0) We didn't consider though that this can happen on remove quote token / remove collateral, need to be added as per your test.

Comment (godsflaw) [reply tests/forge/regression/ERC20Pool/RegressionTestBasicERC20Pool.t.sol:815]: We've been working this up all weekend into an attack, and your fix was one of my suggestions, but  more generally: yes, we think there could be a more systemic issue of rounding in favor of the user in other areas too, and finding and fixing those cases may be the better path than patching each one. We'll write it up in the final report.

Comment (grandizzy) [reply tests/forge/regression/ERC20Pool/RegressionTestBasicERC20Pool.t.sol:815]: thank you!

File: tests/forge/regression/ERC20Pool/RegressionTestReservesERC20Pool.t.sol [hunk 4]
```diff
@@ -959,4 +959,90 @@ contract RegressionTestReserveWith10BucketsERC20Pool is ReserveERC20PoolInvarian
...
+        _reserveERC20PoolHandler.takeAuction(2, 115792089237316195423570985008687907853269984665640564039457584007913129639932, 247429948148584627149392386958522289920829622687852166342, 115792089237316195423570985008687907853269984665640564039457584007913129639934);
+        _reserveERC20PoolHandler.takeReserves(361466280750769417526873054117455902300441173869322632522537, 776804005938781782558157697775153837076826088268880722908914, 115792089237316195423570985008687907853269984665640564039457584007913129639934);
+        _reserveERC20PoolHandler.drawDebt(139839408536053689189893101355857, 1, 10528164681253760395808876505656603171974);
+        _reserveERC20PoolHandler.kickAuction(3589583182297393115565481914547, 41668, 1380496931690591613721258081, 14181);
+        _reserveERC20PoolHandler.drawDebt(9014812045836651946560288041872, 38936, 6000187733701002610517000054944);
+
+        _invariant_B5_B6_B7();
+    }
+
+    function test_regression_prototech_R8() external {
```
Comment (godsflaw) [inline tests/forge/regression/ERC20Pool/RegressionTestReservesERC20Pool.t.sol:1181]: okay, so this invariant test failed:
```solidity
                    // Common case, 1 one millionth (0.000_001) of a quote token or greater is inserted into a single bucket
                    requireWithinDiff(
                        currentExchangeRate,
                        previousExchangeRate,
                        1e13,  // otherwise require exchange rates to be within 1e-5
                        "Exchange Rate Invariant R1, R2, R3, R4, R5, R6, R7 or R8"
                    );
```
because:
```
  ======================================
  Bucket Index           --> 2587
  Previous exchange Rate --> 1000000000000000001
  Current exchange Rate  --> 999980704607586875
  Current bucket lps     --> 15890612102926550116
  ======================================
```
so this does a `1000000000000000001 - 999980704607586875 = 19295392413126`
so in WAD: `19295392413126 / 10^18 = .000019295392413126` 
ultimately, the rate change was greater than `1e13` at:
``` 
>>> 19295392413126 / 10^18
.000019295392413126
>>> 10^13/10^18
.000010000000000000
```
Not sure if the chosen threshold is arbitrarily or if this is bound by something in the spec.  If the later, then this is an issue.  I have my debug printing in still, and the remove collateral did not trigger the rounding edge case.  So this may be a different issue than the last regression.  I'll come back to this unless someone else wants to dig in.

File: tests/forge/regression/ERC20Pool/RegressionTestBasicERC20Pool.t.sol [hunk 5]
```diff
@@ -466,3 +484,63 @@ contract RegressionTestBasicWith10BucketsERC20Pool is BasicERC20PoolInvariants {
...
+contract RegressionTestBasicWith50BucketsIndex500ERC20Pool is BasicERC20PoolInvariants {
+
+    function setUp() public override {
+        // failures reproduced with 50 active buckets
+        vm.setEnv("NO_OF_BUCKETS", "50");
+        vm.setEnv("BUCKET_INDEX_ERC20", "500");
+        super.setUp();
+    }
+
+    function test_regression_prototech_R9() external {
```
Comment (godsflaw) [inline tests/forge/regression/ERC20Pool/RegressionTestBasicERC20Pool.t.sol:850]: This is the most interesting of the three recent rate invariant regressions (R8, R9, R10) as the rate here is not just a little off the threshold when tested against the invariant, but extremely off, meaning it's probably a valid invariant violation.
```
>>> (1020110026839818764 - 1000000000000000001)/10^18
.020110026839818763
```

File: tests/forge/unit/ERC20Pool/ERC20PoolCollateral.t.sol [hunk 6]
```diff
@@ -1044,4 +1044,119 @@ contract ERC20PoolCollateralTest is ERC20HelperContract {
         // borrower should be able to repay and pull collateral
         ERC20Pool(address(_pool)).repayDebt(actor, 120, 1, actor, 7388);
     }
+
+    function test_prototech_collateral_draining() external {
```
Comment (godsflaw) [inline tests/forge/unit/ERC20Pool/ERC20PoolCollateral.t.sol]: This is still a work-in-progress.

File: tests/forge/regression/ERC20Pool/RegressionTestBasicERC20Pool.t.sol (hunk 5 above)
Comment (kmbarry1) [reply tests/forge/regression/ERC20Pool/RegressionTestBasicERC20Pool.t.sol:850]: Nice, this looks promising, will spend the afternoon/evening digging into it.

Comment (kmbarry1) [reply tests/forge/regression/ERC20Pool/RegressionTestBasicERC20Pool.t.sol:850]: Hmmmm this one is decreasing, not increasing, for me:
```
  Previous exchange Rate --> 1000000000000000019
  Current exchange Rate  --> 990238875394222620
```

Comment (godsflaw) [reply tests/forge/regression/ERC20Pool/RegressionTestBasicERC20Pool.t.sol:850]: There is some change introduced between the audit version and this that I need to track down.

Comment (godsflaw) [reply tests/forge/regression/ERC20Pool/RegressionTestBasicERC20Pool.t.sol:850]: The change responsible for the different values happened at https://github.com/ajna-finance/contracts/commit/5f2f5288a38ef8af1d94046c525f268c0f0c11dc in the change to `boundedAmount_`.  You can play with either version with the following patch on that commit, then toggle the comment on `boundedAmount_`.  Save this patch to the file `test_regression_prototech_R9.patch` and run `git apply test_regression_prototech_R9.patch` on a checkout of that commit hash.
```diff
diff --git a/tests/forge/invariants/ERC20Pool/handlers/BasicERC20PoolHandler.sol b/tests/forge/invariants/ERC20Pool/handlers/BasicERC20PoolHandler.sol
index fc65d4a4..39bfa1ed 100644
--- a/tests/forge/invariants/ERC20Pool/handlers/BasicERC20PoolHandler.sol
+++ b/tests/forge/invariants/ERC20Pool/handlers/BasicERC20PoolHandler.sol
@@ -169,6 +169,7 @@ contract BasicERC20PoolHandler is UnboundedBasicERC20PoolHandler, BasicPoolHandl
         uint256 amountToBorrow_
     ) internal override returns (uint256 boundedAmount_) {
         boundedAmount_ = constrictToRange(amountToBorrow_, MIN_QUOTE_AMOUNT, MAX_QUOTE_AMOUNT);
+        // boundedAmount_ = constrictToRange(amountToBorrow_, MIN_DEBT_AMOUNT, MAX_DEBT_AMOUNT);
 
         // Pre Condition
         // 1. borrower's debt should exceed minDebt
diff --git a/tests/forge/invariants/ERC20Pool/handlers/unbounded/BaseERC20PoolHandler.sol b/tests/forge/invariants/ERC20Pool/handlers/unbounded/BaseERC20PoolHandler.sol
index 44d176a5..836a06d5 100644
--- a/tests/forge/invariants/ERC20Pool/handlers/unbounded/BaseERC20PoolHandler.sol
+++ b/tests/forge/invariants/ERC20Pool/handlers/unbounded/BaseERC20PoolHandler.sol
@@ -37,6 +37,9 @@ abstract contract BaseERC20PoolHandler is BaseHandler {
         MIN_QUOTE_AMOUNT = vm.envOr("MIN_QUOTE_AMOUNT_ERC20", uint256(1e3));
         MAX_QUOTE_AMOUNT = vm.envOr("MAX_QUOTE_AMOUNT_ERC20", uint256(1e30));
 
+        MIN_DEBT_AMOUNT = vm.envOr("MIN_DEBT_AMOUNT", uint256(0));
+        MAX_DEBT_AMOUNT = vm.envOr("MAX_DEBT_AMOUNT", uint256(1e28));
+
         MIN_COLLATERAL_AMOUNT = vm.envOr("MIN_COLLATERAL_AMOUNT_ERC20", uint256(1e3));
         MAX_COLLATERAL_AMOUNT = vm.envOr("MAX_COLLATERAL_AMOUNT_ERC20", uint256(1e30));
 
@@ -77,4 +80,4 @@ abstract contract BaseERC20PoolHandler is BaseHandler {
         return actorsAddress;
     }
 
-}
\ No newline at end of file
+}
diff --git a/tests/forge/invariants/base/handlers/unbounded/BaseHandler.sol b/tests/forge/invariants/base/handlers/unbounded/BaseHandler.sol
index a6349d3d..9ae68aa7 100644
--- a/tests/forge/invariants/base/handlers/unbounded/BaseHandler.sol
+++ b/tests/forge/invariants/base/handlers/unbounded/BaseHandler.sol
@@ -40,6 +40,8 @@ abstract contract BaseHandler is Test {
 
     uint256 internal MIN_QUOTE_AMOUNT;
     uint256 internal MAX_QUOTE_AMOUNT;
+    uint256 internal MIN_DEBT_AMOUNT;
+    uint256 internal MAX_DEBT_AMOUNT;
 
     uint256 internal MIN_COLLATERAL_AMOUNT;
     uint256 internal MAX_COLLATERAL_AMOUNT;
diff --git a/tests/forge/regression/ERC20Pool/RegressionTestBasicERC20Pool.t.sol b/tests/forge/regression/ERC20Pool/RegressionTestBasicERC20Pool.t.sol
index e0570a7e..e7ace5d3 100644
--- a/tests/forge/regression/ERC20Pool/RegressionTestBasicERC20Pool.t.sol
+++ b/tests/forge/regression/ERC20Pool/RegressionTestBasicERC20Pool.t.sol
@@ -377,3 +377,28 @@ contract RegressionTestBasicWith10BucketsERC20Pool is BasicERC20PoolInvariants {
     }
 }
 
+contract RegressionTestBasicWith50BucketsIndex500ERC20Pool is BasicERC20PoolInvariants {
+
+    function setUp() public override {
+        // failures reproduced with 50 active buckets
+        vm.setEnv("NO_OF_BUCKETS", "50");
+        vm.setEnv("BUCKET_INDEX_ERC20", "500");
+        super.setUp();
+    }
+
+    function test_regression_prototech_R9() external {
+        _basicERC20PoolHandler.removeCollateral(115792089237316195423570985008687907853269984665640564039457584007913129639932, 3, 115792089237316195423570985008687907853269984665640564039457584007913129639933, 311855305968157744570601481773144);
+        _basicERC20PoolHandler.addCollateral(2, 115792089237316195423570985008687907853269984665640564039457584007913129639933, 0, 115792089237316195423570985008687907853269984665640564039457584007913129639933);
+        _basicERC20PoolHandler.removeCollateral(169050319556909991808, 3, 0, 1);
+        _basicERC20PoolHandler.addCollateral(1004201281585420667067519110248, 7102026478996272106877444035857, 1000005176609260307609491, 3305068122278678061826709665672);
+        _basicERC20PoolHandler.removeCollateral(5126250508003894776370908264788, 491360638543797591635259813352, 787983051556723253156194660814, 3320088216653581211669326645577);
+        _basicERC20PoolHandler.repayDebt(12396923902833375051988245672959, 7853833558731273234852069656215, 4854998063651887910481309782641);
+        _basicERC20PoolHandler.addQuoteToken(22492324757351622016450338305141639519921418025875019955286375595978, 7375145817336195198480525641381057, 4038098431544972508850862464533, 590641590);
+        _basicERC20PoolHandler.moveQuoteToken(2, 18642700475706059316512354602605761623058115872316, 0, 389852391755003176683454195550428602558533101523111398325617912, 115792089237316195423570985008687907853269984665640564039457584007913129639932);
+        _basicERC20PoolHandler.removeCollateral(183508, 4655513824786435541548325526158, 6727168359250653708446790113706, 689384358901091345705449093185695413140861121050633943735045187092801409);
+        _basicERC20PoolHandler.removeCollateral(887908587686943510167377721632, 459811254525365132505830027215, 4833907401013214004421139781796, 353362243998154774896346551976);
+        _basicERC20PoolHandler.removeCollateral(2, 115792089237316195423570985008687907853269984665640564039457584007913129639934, 12898304536022767664608434174578363454046958904043510776692055820203777739, 115792089237316195423570985008687907853269984665640564039457584007913129639932);
+
+        invariant_rate_R1_R2_R3_R4_R5_R6_R7_R8();
+    }
+}

```

File: tests/forge/regression/ERC20Pool/RegressionTestLiquidationERC20Pool.t.sol [hunk 7]
```diff
@@ -1970,4 +1972,272 @@ contract RegressionTestLiquidationWith20BucketsERC20Pool is LiquidationERC20Pool
...
+
+contract RegressionTestLiquidationWith50BucketsIndex1ERC20Pool is LiquidationERC20PoolInvariants {
+
+    function setUp() public override {
+        vm.setEnv("NO_OF_BUCKETS", "50");
+        vm.setEnv("BUCKET_INDEX_ERC20", "1");
+        super.setUp();
+    }
+    
+    function test_regression_prototech_R14() external {
```
Comment (godsflaw) [inline tests/forge/regression/ERC20Pool/RegressionTestLiquidationERC20Pool.t.sol:2081]: This one is different than most of the exchange rate invariant violations:
```rust
Running 1 test for tests/forge/regression/ERC20Pool/RegressionTestLiquidationERC20Pool.t.sol:RegressionTestLiquidationWith50BucketsIndex1ERC20Pool
[FAIL. Reason: Buckets Invariant B2] test_regression_prototech_R14() (gas: 723496045)
Test result: FAILED. 0 passed; 1 failed; finished in 2.93s
```

From this code:
```solidity
            uint256 bucketIndex = buckets[i];
            (
                ,
                uint256 deposit,
                uint256 collateral,
                uint256 bucketLps,
                ,
                uint256 exchangeRate
            ) = _poolInfo.bucketInfo(address(_pool), bucketIndex);

            if (collateral == 0 && deposit == 0) {
                require(bucketLps == 0,       "Buckets Invariant B2");
                require(exchangeRate == 1e18, "Buckets Invariant B3");
            }
```

which we can see from the trace:
```rust
    ├─ [21430] PoolInfoUtils::bucketInfo(0x9Fd7552Da37D2D296CB4a84d507c35Dcc926220a, 2) [staticcall]
    │   ├─ [9757] 0x9Fd7552Da37D2D296CB4a84d507c35Dcc926220a::bucketInfo(2) [staticcall]
    │   │   ├─ [9536] ERC20Pool::bucketInfo(2) [delegatecall]
    │   │   │   └─ ← 299060495, 0, 0, 0, 604913114103258654822514985
    │   │   └─ ← 299060495, 0, 0, 0, 604913114103258654822514985
    │   └─ ← 994994171041818143490934622, 0, 0, 299060495, 604913114103258654822514985, 0
    └─ ← "Buckets Invariant B2"
```
That we are left with `299060495` LP in the bucket 😱 .

This is bucket `2`, so we are again on the extremes of price and are likely hitting another rounding edge case at those extremes.

Comment (grandizzy) [reply tests/forge/regression/ERC20Pool/RegressionTestLiquidationERC20Pool.t.sol:2081]: this happens due to adding an amount in bucket 2 that rounded is 0, so one could end up with adding no quote tokens in pool but getting LP out of it 
```
  awardedLP: 299060495
  scaled amount to add: 299060495
  bucket scale: 604913114680856126342163304
  unscaled amount added(Maths.wdiv(scaled amount to add, bucket scale)): 0
```
 we only revert if `(bucketLP_ == 0) revert InsufficientLP();` but the amount added is not checked.

Comment (godsflaw) [reply tests/forge/regression/ERC20Pool/RegressionTestLiquidationERC20Pool.t.sol:2081]: Looks like you added this fix here: https://github.com/ajna-finance/contracts/pull/841

**Comment by godsflaw** [comment]
After rebasing this PR onto https://github.com/ajna-finance/contracts/pull/841 I still get a few regression failures listed below:

```rust
Failing tests:
Encountered 1 failing test in tests/forge/regression/ERC20Pool/RegressionTestBasicERC20Pool.t.sol:RegressionTestBasicERC20Pool
[FAIL. Reason: Assertion failed.] test_regression_fenwick_index_2() (gas: 8686398)

Encountered 1 failing test in tests/forge/regression/ERC20Pool/RegressionTestReservesERC20Pool.t.sol:RegressionTestReserveWith50BucketsERC20Pool
[FAIL. Reason: Exchange Rate Invariant R1, R2, R3, R4, R5, R6, R7 or R8] test_regression_prototech_R12() (gas: 215843752)

Encountered 1 failing test in tests/forge/regression/ERC20Pool/RegressionTestReservesERC20Pool.t.sol:RegressionTestReserveWith50BucketsIndex1ERC20Pool
[FAIL. Reason: Exchange Rate Invariant R1, R2, R3, R4, R5, R6, R7 or R8] test_regression_prototech_R10() (gas: 171578553)

Encountered 1 failing test in tests/forge/regression/ERC20Pool/RegressionTestReservesERC20Pool.t.sol:RegressionTestReserveWith50BucketsIndex500ERC20Pool
[FAIL. Reason: Exchange Rate Invariant R1, R2, R3, R4, R5, R6, R7 or R8] test_regression_prototech_R17() (gas: 219256425)
```

File: tests/forge/regression/ERC20Pool/RegressionTestLiquidationERC20Pool.t.sol (hunk 7 above)
Comment (godsflaw) [reply tests/forge/regression/ERC20Pool/RegressionTestLiquidationERC20Pool.t.sol:2081]: This fix did indeed resolve the issue after merge.

File: tests/forge/regression/ERC20Pool/RegressionTestBasicERC20Pool.t.sol (hunk 5 above)
Comment (godsflaw) [reply tests/forge/regression/ERC20Pool/RegressionTestBasicERC20Pool.t.sol:850]: issue resolved in: https://github.com/ajna-finance/contracts/pull/841

File: tests/forge/regression/ERC20Pool/RegressionTestReservesERC20Pool.t.sol (hunk 4 above)
Comment (godsflaw) [reply tests/forge/regression/ERC20Pool/RegressionTestReservesERC20Pool.t.sol:1181]: issue resolved in https://github.com/ajna-finance/contracts/pull/841

File: tests/forge/regression/ERC20Pool/RegressionTestBasicERC20Pool.t.sol (hunk 3 above)
Comment (godsflaw) [reply tests/forge/regression/ERC20Pool/RegressionTestBasicERC20Pool.t.sol:815]: issue resolve in https://github.com/ajna-finance/contracts/pull/841
